File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/ui/login/LoginViewModel.kt
package com.barkfluff.BarkCloud.ui.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.CreationExtras
import com.barkfluff.BarkCloud.BarkCloudApplication
import com.barkfluff.BarkCloud.data.AuthRepository
import com.barkfluff.BarkCloud.data.AuthResult
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class LoginViewModel(
    private val authRepository: AuthRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(LoginUiState())
    val state: StateFlow<LoginUiState> = _state.asStateFlow()

    private val _events = Channel<LoginEvent>(Channel.BUFFERED)
    val events = _events.receiveAsFlow()

    fun onLoginChange(value: String) {
        _state.update { it.copy(login = value, credentialsError = null) }
    }

    fun onPasswordChange(value: String) {
        _state.update { it.copy(password = value, credentialsError = null) }
    }

    fun onPasswordVisibilityToggle() {
        _state.update { it.copy(passwordVisible = !it.passwordVisible) }
    }

    fun onOtpChange(value: String) {
        val sanitized = value.filter { it.isDigit() }.take(LoginUiState.OTP_LENGTH)
        _state.update { it.copy(otp = sanitized) }
        if (sanitized.length == LoginUiState.OTP_LENGTH) {
            submit()
        }
    }

    fun onComingSoon() {
        _state.update { it.copy(snackbarMessage = COMING_SOON) }
    }

    fun snackbarShown() {
        _state.update { it.copy(snackbarMessage = null) }
    }

    fun submit() {
        val current = _state.value
        if (!current.canSubmit) return
        _state.update { it.copy(isLoading = true, credentialsError = null) }

        viewModelScope.launch {
            val result = authRepository.auth(
                login = current.login.trim(),
                password = current.password,
                otpCode = current.otp.takeIf { current.otpRequired },
            )
            when (result) {
                AuthResult.Success -> {
                    _state.update { it.copy(isLoading = false) }
                    _events.send(LoginEvent.NavigateToMain)
                }
                AuthResult.OtpRequired -> {
                    _state.update {
                        it.copy(
                            isLoading = false,
                            otpRequired = true,
                            otp = "",
                        )
                    }
                }
                AuthResult.InvalidCredentials -> {
                    _state.update {
                        it.copy(
                            isLoading = false,
                            credentialsError = INVALID_CREDENTIALS,
                        )
                    }
                }
                is AuthResult.OtherError -> {
                    _state.update {
                        it.copy(
                            isLoading = false,
                            snackbarMessage = result.message.ifBlank { NETWORK_ERROR },
                        )
                    }
                }
            }
        }
    }

    sealed class LoginEvent {
        data object NavigateToMain : LoginEvent()
    }

    companion object {
        private const val COMING_SOON = "Скоро"
        private const val INVALID_CREDENTIALS = "Неверный логин или пароль"
        private const val NETWORK_ERROR = "Не удалось связаться с сервером"

        fun factory(): ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
                val app = extras[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY]
                    as BarkCloudApplication
                return LoginViewModel(app.authRepository) as T
            }
        }
    }
}

File: Ios/BarkCloud/BarkCloud/Features/Login/LoginViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class LoginViewModel {
    var state = LoginUiState()
    var navigateToMainRequest: Bool = false

    private let auth: AuthRepository

    init(auth: AuthRepository) {
        self.auth = auth
    }

    func onLoginChange(_ value: String) {
        state.login = value
        state.credentialsError = nil
    }

    func onPasswordChange(_ value: String) {
        state.password = value
        state.credentialsError = nil
    }

    func togglePasswordVisibility() {
        state.passwordVisible.toggle()
    }

    func onOtpChange(_ value: String) {
        let filtered = String(value.filter(\.isNumber).prefix(LoginUiState.otpLength))
        state.otp = filtered
        if filtered.count == LoginUiState.otpLength {
            submit()
        }
    }

    func onComingSoon() {
        state.snackbarMessage = String(localized: "login_coming_soon")
    }

    func snackbarShown() {
        state.snackbarMessage = nil
    }

    func submit() {
        guard state.canSubmit else { return }
        state.isLoading = true
        state.credentialsError = nil
        let loginValue = state.login.trimmingCharacters(in: .whitespaces)
        let passwordValue = state.password
        let otpValue: String? = state.otpRequired ? state.otp : nil

        Task { [auth] in
            let result = await auth.auth(login: loginValue, password: passwordValue, otpCode: otpValue)
            await MainActor.run {
                self.handle(result)
            }
        }
    }

    private func handle(_ result: AuthResult) {
        state.isLoading = false
        switch result {
        case .success:
            navigateToMainRequest = true
        case .otpRequired:
            state.otpRequired = true
            state.otp = ""
        case .invalidCredentials:
            state.credentialsError = String(localized: "login_error_invalid_credentials")
        case .otherError(let message):
            state.snackbarMessage = message.isEmpty ? String(localized: "login_error_network") : message
        }
    }
}
